annotation-solr/src/main/java/eu/europeana/annotation/solr/service/impl/SolrAnnotationServiceImpl.java

```diff
@@ -539,87 +538,90 @@ else if(BodyInternalTypes.isSimpleTagBody(anno.getBody().getInternalType()))
-		List<String> targetValues = anno.getTarget().getValues();
-		for (String targetValue : targetValues) {
-			solrQuery+=" AND " + SolrAnnotationConstants.TARGET_URI + ":\"" + targetValue + "\"";
-		}
-		solrQuery+=")";
-		return solrQuery;
+	private SolrQuery solrUniquenessQuerySimpleTagging(Annotation anno) {
+
+      SolrQuery query = new SolrQuery();
+      query.setQuery(SolrAnnotationConstants.TARGET_URI + ":\"" + anno.getTarget().getValue() + "\"");
+      query.addFilterQuery(WebAnnotationModelFields.MOTIVATION + ":\"" + MotivationTypes.TAGGING.getOaType() + "\"");
+      query.addFilterQuery(SolrAnnotationConstants.BODY_VALUE + ":\"" + anno.getBody().getValue() + "\"");
+      return query;
 	}
 	
-	private String solrUniquenessQueryLinkForContributing(Annotation anno) {
-      String solrQuery = "(" + WebAnnotationModelFields.MOTIVATION + ":\"" + MotivationTypes.LINKFORCONTRIBUTING.getOaType() + "\"";
-      
+	private SolrQuery solrUniquenessQueryLinking(Annotation anno) {
+      SolrQuery query = new SolrQuery();
       List<String> targetValues = anno.getTarget().getValues();
       if(targetValues!=null) {
-        for (String targetValue : targetValues) {
-            solrQuery+=" AND " + SolrAnnotationConstants.TARGET_URI + ":\"" + targetValue + "\"";
+        String targetUrisQuery = "";
+        for (int i=0; i<targetValues.size(); i++) {
+          if(i==0)
+            targetUrisQuery += SolrAnnotationConstants.TARGET_URI + ":\"" + targetValues.get(i) + "\"";
+          else
+            targetUrisQuery +=" AND " + SolrAnnotationConstants.TARGET_URI + ":\"" + targetValues.get(i) + "\"";
         }
+        query.setQuery(targetUrisQuery);
       }
+      else
+        query.setQuery("*:*");
       
-      if(anno.getTarget().getValue()!=null) {
-        solrQuery+=" AND " + SolrAnnotationConstants.TARGET_URI + ":\"" + anno.getTarget().getValue() + "\"";
-      }
-
+      query.addFilterQuery(WebAnnotationModelFields.MOTIVATION + ":\"" + MotivationTypes.LINKING.getOaType() + "\"");
+      return query;	  
+	}
+	
+	private SolrQuery solrUniquenessQueryLinkForContributing(Annotation anno) {
+      SolrQuery query = new SolrQuery();
+      query.setQuery(SolrAnnotationConstants.TARGET_URI + ":\"" + anno.getTarget().getValue() + "\"");
+      query.addFilterQuery(WebAnnotationModelFields.MOTIVATION + ":\"" + MotivationTypes.LINKFORCONTRIBUTING.getOaType() + "\"");
       List<String> bodyUris = extractUriValues(anno.getBody());
-      if(bodyUris!=null) {
-        for (String bodyUri : bodyUris) {
-            solrQuery+=" AND " + SolrAnnotationConstants.BODY_URI + ":\"" + bodyUri + "\""; 
-        }
+      //using a single filter here (see please: https://solr.apache.org/guide/8_11/common-query-parameters.html#fq-filter-query-parameter, for more information)
+      String bodyUrisFilter = "";
+      for (int i=0; i<bodyUris.size(); i++) {
```

Comment:
for maintenance purposes, it is better to not concatenate URIs in the query but simply add new filter for each URI

Reply:
Updated.